### PR TITLE
feat(firebase_messaging, android)!: android 13 notifications permission request

### PR DESCRIPTION
**Pull request by russellwheatley**
## Description

PR checks if the app is running on android 13 and will check the app notification permissions. If none set or permissions not granted, a dialog box will appear requesting user notification permission.

I consider this a breaking change as there is certain configuration required such as `targetSdkVersion` & `compileSdkVersion` targeting android API 33. 

Users will also need to have the `android:exported="true"` property set in app manifest on apps running on android 12 or higher.

## Related Issues

closes https://github.com/firebase/flutterfire/issues/8720

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`).
This will ensure a smooth and quick review process. Updating the `pubspec.yaml` and changelogs is not required.

- [ ] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [ ] My PR includes unit or integration tests for *all* changed/updated/fixed behaviors (See [Contributor Guide]).
- [ ] All existing and new tests are passing.
- [ ] I updated/added relevant documentation (doc comments with `///`).
- [ ] The analyzer (`melos run analyze`) does not report any problems on my PR.
- [ ] I read and followed the [Flutter Style Guide].
- [ ] I signed the [CLA].
- [ ] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Does your PR require plugin users to manually update their apps to accommodate your change?

- [X] Yes, this is a breaking change.
- [] No, this is *not* a breaking change.

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/firebase/flutterfire/blob/master/CONTRIBUTING.md
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[pub versioning philosophy]: https://dart.dev/tools/pub/versioning
[CLA]: https://cla.developers.google.com/

### CODE REVIEW

**Comment by MarcinusX** [comment]
@russellwheatley According to [msackman's comment](https://github.com/firebase/flutterfire/issues/8720#issuecomment-1216454840), I think we also need to update `getPermissions()` method to check for:

```java
final boolean areNotificationsEnabled = ContextHolder.getApplicationContext().checkSelfPermission(Manifest.permission.POST_NOTIFICATIONS) == PackageManager.PERMISSION_GRANTED;
```

**Comment by russellwheatley** [comment]
@MarcinusX That makes sense. I'll update the PR 👍

File: packages/firebase_messaging/firebase_messaging/android/src/main/AndroidManifest.xml [hunk 1]
```diff
@@ -3,6 +3,8 @@
   <uses-permission android:name="android.permission.INTERNET"/>
   <uses-permission android:name="android.permission.WAKE_LOCK"/>
   <uses-permission android:name="android.permission.ACCESS_NETWORK_STATE"/>
+  <!-- Permissions options for the `notification` group -->
+  <uses-permission android:name="android.permission.POST_NOTIFICATIONS"/>
```
Comment (Lyokone) [inline packages/firebase_messaging/firebase_messaging/android/src/main/AndroidManifest.xml:7]: We should probably mention this in the changelog for this PR

File: packages/firebase_messaging/firebase_messaging/android/src/main/java/io/flutter/plugins/firebase/messaging/FlutterFirebaseMessagingPlugin.java [hunk 2]
```diff
@@ -72,14 +77,14 @@ public void onAttachedToEngine(FlutterPluginBinding binding) {
 
   @Override
   public void onDetachedFromEngine(@NonNull FlutterPluginBinding binding) {
-    if (binding.getApplicationContext() != null) {
-      LocalBroadcastManager.getInstance(binding.getApplicationContext()).unregisterReceiver(this);
-    }
+    binding.getApplicationContext();
```
Comment (Lyokone) [inline packages/firebase_messaging/firebase_messaging/android/src/main/java/io/flutter/plugins/firebase/messaging/FlutterFirebaseMessagingPlugin.java:80]: This line doesn't seem to do anything

File: packages/firebase_messaging/firebase_messaging/android/src/main/java/io/flutter/plugins/firebase/messaging/FlutterFirebaseMessagingPlugin.java [hunk 3]
```diff
@@ -311,16 +316,59 @@ private Task<Map<String, Object>> getInitialMessage(Map<String, Object> argument
...
+  private Task<Map<String, Integer>> requestPermissions() {
+    TaskCompletionSource<Map<String, Integer>> taskCompletionSource = new TaskCompletionSource<>();
+    cachedThreadPool.execute(
+        () -> {
+          final Map<String, Integer> permissions = new HashMap<>();
+          try {
+            final boolean areNotificationsEnabled =
+                ContextHolder.getApplicationContext()
+                        .checkSelfPermission(Manifest.permission.POST_NOTIFICATIONS)
+                    == PackageManager.PERMISSION_GRANTED;
```
Comment (Lyokone) [inline packages/firebase_messaging/firebase_messaging/android/src/main/java/io/flutter/plugins/firebase/messaging/FlutterFirebaseMessagingPlugin.java:329]: We repeat 2 times this command, should it be extracted to another function?

File: packages/firebase_messaging/firebase_messaging/android/src/main/java/io/flutter/plugins/firebase/messaging/FlutterFirebaseMessagingPlugin.java [hunk 4]
```diff
@@ -311,16 +316,59 @@ private Task<Map<String, Object>> getInitialMessage(Map<String, Object> argument
...
           try {
             final Map<String, Integer> permissions = new HashMap<>();
-            final boolean areNotificationsEnabled =
-                NotificationManagerCompat.from(mainActivity).areNotificationsEnabled();
-            permissions.put("authorizationStatus", areNotificationsEnabled ? 1 : 0);
+            if (Build.VERSION.SDK_INT >= 33) {
+              final boolean areNotificationsEnabled =
+                  ContextHolder.getApplicationContext()
+                          .checkSelfPermission(Manifest.permission.POST_NOTIFICATIONS)
+                      == PackageManager.PERMISSION_GRANTED;
```
Comment (Lyokone) [inline packages/firebase_messaging/firebase_messaging/android/src/main/java/io/flutter/plugins/firebase/messaging/FlutterFirebaseMessagingPlugin.java:365]: Second time

**Review by Lyokone — CHANGES_REQUESTED** [review]

File: packages/firebase_messaging/firebase_messaging/android/src/main/AndroidManifest.xml (hunk 1 above)
Comment (russellwheatley) [reply packages/firebase_messaging/firebase_messaging/android/src/main/AndroidManifest.xml:7]: We can manually edit this when doing the next release I guess 🤔

Comment (Lyokone) [reply packages/firebase_messaging/firebase_messaging/android/src/main/AndroidManifest.xml:7]: I think you can change the squash message when merging from GitHub directly

**Review by Lyokone — APPROVED** [review]

File: packages/firebase_messaging/firebase_messaging/android/src/main/java/io/flutter/plugins/firebase/messaging/FlutterFirebaseMessagingPlugin.java [hunk 5]
```diff
@@ -311,16 +315,64 @@ private Task<Map<String, Object>> getInitialMessage(Map<String, Object> argument
...
+  }
+
+  private Boolean checkPermissions() {
+    if (Build.VERSION.SDK_INT >= 33) {
+      return ContextHolder.getApplicationContext()
+              .checkSelfPermission(Manifest.permission.POST_NOTIFICATIONS)
+          == PackageManager.PERMISSION_GRANTED;
+    }
+
+    throw new RuntimeException(
```
Comment (Ehesp) [inline packages/firebase_messaging/firebase_messaging/android/src/main/java/io/flutter/plugins/firebase/messaging/FlutterFirebaseMessagingPlugin.java:357]: How is this handled in Dart land? What error code is returned?

**Review by Salakar — APPROVED** [review]

**Pull request merged**